Add support for parsing regex named capture groups to AST.

SharafMohamed opened the pull request:
# Description
- Initial step for implementing parsing with capture groups.
- Parses capture groups in schema file.
- Currently only adds capture group to AST, but doesn't add to NFA/DFA.
- Move initialization of regex_special_characters into cpp to make it more clear what the lexical rules are for the schema parser.
- Format finite automata files.

# Validation performed
- Ensure the newly added AST doesn't break the existing demo in examples.
- Add unit tests for capture groups.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -420,6 +420,57 @@ class RegexASTMultiplication : public RegexAST<NFAStateType> {
+    auto add(RegexNFA<NFAStateType>* nfa, NFAStateType* end_state) -> void override;
+
+private:
+    std::string m_name;
```
How about `m_group_name` since it's the name of the captured group?

SharafMohamed replied:
Changed.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -420,6 +420,57 @@ class RegexASTMultiplication : public RegexAST<NFAStateType> {
 };
+
+template <typename NFAStateType>
+class RegexASTCapture : public RegexAST<NFAStateType> {
```
Do we need a method to get the underlying name?

SharafMohamed replied:
I don't think so. What we need to make this class useful is for the add function to do more than just connect the NFA to the end_state, which is essentially a null operation. This requires the tagged DFA PR, so for now not much can be done.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -420,6 +420,57 @@ class RegexASTMultiplication : public RegexAST<NFAStateType> {
+class RegexASTCapture : public RegexAST<NFAStateType> {
+public:
+    RegexASTCapture(
+            std::string /*name*/,
+            std::unique_ptr<RegexAST<NFAStateType>> /*regex*/
```
How about moving the implementation from tpp to here?
Even if we don't, we should complete the signature with `name` and `regex` explicitly named instead of using comments

SharafMohamed replied:
All the classes inheriting from RegexAST currently define their constructors in the tpp. For now going to leave it in the tpp for consistency, even though I agree it should be in the header. Will do a following PR to focus on fixing some stylistic issues that applies to more than just the code touched here. 

Changed signature to be explicit.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.tpp:
```diff
@@ -116,18 +116,31 @@ void RegexASTMultiplication<NFAStateType>::add(
+        std::string name,
+        std::unique_ptr<RegexAST<NFAStateType>> regex
+)
+        : m_name(name),
```
```suggestion
        : m_name(std::move(name)),
```

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -420,6 +420,57 @@ class RegexASTMultiplication : public RegexAST<NFAStateType> {
+    }
+
+    /**
+     * Sets is_possible_input to specify which utf8 characters are allowed in a
+     * lexer rule containing RegexASTCat at a leaf node in its AST
```
```suggestion
     * Sets `is_possible_input` to specify which utf8 characters are allowed in a
     * lexer rule containing `RegexASTCat` at a leaf node in its AST
```

SharafMohamed replied:
Updated.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -420,6 +420,57 @@ class RegexASTMultiplication : public RegexAST<NFAStateType> {
+    }
+
+    /**
+     * Add the needed RegexNFA::states to the passed in nfa to handle a
+     * RegexASTCapture before transitioning to a pre-tagged end_state
```
```suggestion
     * Adds the needed `RegexNFA::states` to the passed in nfa to handle a
     * `RegexASTCapture` before transitioning to a pre-tagged `end_state`
```

SharafMohamed replied:
Changed and applied same idea to other comments in RegexASTCapture class.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -420,6 +420,57 @@ class RegexASTMultiplication : public RegexAST<NFAStateType> {
+
+private:
+    std::string m_name;
+    std::unique_ptr<RegexAST<NFAStateType>> m_regex;
```
`m_regex` is a bit confusing, since the class is also called `RegexXXX`. Maybe `m_pattern` or sth indicating that it's a sub AST of the pattern to capture?

SharafMohamed replied:
Its not a subAST, its the entire AST for the capture group. What about m_group_regex_ast, as it is a RegexAST?

LinZhihao-723 replied:
yeah make sense. `m_group_regex_ast` sounds better

LinZhihao-723 commented on src/log_surgeon/LogParser.cpp:
```diff
@@ -88,6 +88,10 @@ void LogParser::add_rules(std::unique_ptr<SchemaAST> schema_ast) {
             continue;
         }
+        // currently capture groups are not yet supported
+        if (rule->m_name == "capture") {
```
```suggestion
        if ("capture" == rule->m_name) {
```

SharafMohamed replied:
Swapped.

LinZhihao-723 reviewed (changes requested):
For the validation, iiuc the `capture` rule isn't parsed yet right? Do we have more detailed explanation about the validation steps?

SharafMohamed commented:
> For the validation, iiuc the `capture` rule isn't parsed yet right? Do we have more detailed explanation about the validation steps?

For validation, we updated the schema.txt in the examples to cancel newly added special characters and added a capture group variable. The capture group variable is only used to create the schema AST, and is currently ignored during log parsing. So the current validation only shows that the capture group change (inclusion of capture groups in the schema.txt and schema AST) doesn't break any existing functionality. 

Currently LogSurgeon has no unit-tests, so there is actually no guarantee any of the schema AST stuff was previously correct, all we know is that the end-to-end benchmarks are succeeding. The next PR is going to address the unit test issue. So maybe we wait on that PR, and add some unit tests for testing the schema AST. Other than that, the only other way to test it would be after the next few PRs that adds tags to the DFA and then the log parsing change, which isn't the best option.

SharafMohamed commented on src/log_surgeon/SchemaParser.hpp:
```diff
@@ -125,21 +125,7 @@ class SchemaParser : public LALR1Parser<
-            {'}', "Rbrace"},
-            {'|', "Vbar"}
-    };
+    static std::unordered_map<char, std::string> m_special_regex_characters;
```
Added inline back.

SharafMohamed commented on src/log_surgeon/SchemaParser.cpp:
```diff
@@ -26,12 +26,17 @@ using RegexASTOrByte
 using std::unique_ptr;
 
 namespace log_surgeon {
+
+std::unordered_map<char, std::string> SchemaParser::m_special_regex_characters;
+
```
Removed.

SharafMohamed commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -420,6 +420,57 @@ class RegexASTMultiplication : public RegexAST<NFAStateType> {
+            std::unique_ptr<RegexAST<NFAStateType>> /*regex*/
+    );
+
+    RegexASTCapture(RegexASTCapture const& rhs)
```
Whoops, definitely need to `m_name` as well.

LinZhihao-723 commented:
> > For the validation, iiuc the `capture` rule isn't parsed yet right? Do we have more detailed explanation about the validation steps?
> 
> For validation, we updated the schema.txt in the examples to cancel newly added special characters and added a capture group variable. The capture group variable is only used to create the schema AST, and is currently ignored during log parsing. So the current validation only shows that the capture group change (inclusion of capture groups in the schema.txt and schema AST) doesn't break any existing functionality.
> 
> Currently LogSurgeon has no unit-tests, so there is actually no guarantee any of the schema AST stuff was previously correct, all we know is that the end-to-end benchmarks are succeeding. The next PR is going to address the unit test issue. So maybe we wait on that PR, and add some unit tests for testing the schema AST. Other than that, the only other way to test it would be after the next few PRs that adds tags to the DFA and then the log parsing change, which isn't the best option.

I see. For the validation performed, I was confused by the description since my understanding is the same as your reply. I think the validation description can be sth like "Ensure the newly added AST doesn't break the existing demo in examples."

LinZhihao-723 commented on src/log_surgeon/SchemaParser.cpp:
```diff
@@ -26,12 +26,15 @@ using RegexASTOrByte
 using std::unique_ptr;
 
 namespace log_surgeon {
+
```
```suggestion
```

SharafMohamed replied:
I'm going to PR the unit test framework separately, and then add a few unit tests to this PR before we approve it.

LinZhihao-723 reviewed (changes requested):
Final nit fix and then I will approve

SharafMohamed commented:
I'm going to PR the unit test framework separately, and then add a few unit tests to this PR before we approve it.

LinZhihao-723 commented:
> I'm going to PR the unit test framework separately, and then add a few unit tests to this PR so you can review them with this.

I will approve now and pass it to @kirkrodrigues

LinZhihao-723 commented on tests/test-lexer.cpp:
```diff
@@ -1,20 +1,80 @@
+    }
+
+    SECTION("Add a capture variable to schema") {
+        schema.add_variable("capture", "u(?<uID>[0-9]+)", -1);
```
Nit: we can make `"capture"` as a `constexpr std::string_view`

SharafMohamed replied:
What about const std::string so it can be used with add_variable(), unless you think add_variable() should be modified.

LinZhihao-723 reviewed (approved):
The unit test makes sense to me
